### ISSUE TITLE
vb.net

### ISSUE THREAD

**Issue by wizzardmr42** [RazorGenerator/RazorGenerator#70]
I've forked this and I'm trying to have a go at getting it working for .vbhtml (never contributed to open source before, so no guarantees!), but I can't work out what needs doing.  Can anyone point me in the right direction?

**Comment by pranavkm** [comment]
You might have to switch around a bunch of code that assumes that the parser is a CSharp parser (for instance (https://github.com/RazorGenerator/RazorGenerator/blob/master/RazorGenerator.Core.v1/ExportedTransformers/MvcViewTransformer.cs#L71). Also places like these might require switching based on the extension: https://github.com/RazorGenerator/RazorGenerator/blob/master/RazorGenerator.Tooling/BaseCodeGeneratorWithSite.cs#L128-L129

We largely use CodeDOM for code manipulation so most of that should remain unchanged.

**Comment by wizzardmr42** [comment]
I'm getting on quite well with it - think I have done most of the work now and the single file generator seems to be working fine with a basic test.  However, I've been getting the following during the build even before I started making changes - any ideas what is causing it?

7>EXEC : warning : 20 issue(s) found with package 'RazorGenerator.MsBuild'.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\RazorGenerator.Core.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\RazorGenerator.MsBuild.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v1\RazorGenerator.Core.v1.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v1\System.Web.Mvc.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v1\System.Web.Razor.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v1\System.Web.WebPages.Deployment.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v1\System.Web.WebPages.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v1\System.Web.WebPages.Razor.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v2\RazorGenerator.Core.V2.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v2\System.Web.Mvc.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v2\System.Web.Razor.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v2\System.Web.WebPages.Deployment.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v2\System.Web.WebPages.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v2\System.Web.WebPages.Razor.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v3\RazorGenerator.Core.V3.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v3\System.Web.Mvc.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v3\System.Web.Razor.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v3\System.Web.WebPages.Deployment.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v3\System.Web.WebPages.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.
7>  
7>  Issue: Assembly outside lib folder.
7>  Description: The assembly 'tools\v3\System.Web.WebPages.Razor.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project.
7>  Solution: Move it into the 'lib' folder if it should be referenced.

**Comment by pranavkm** [comment]
Feel free to ignore those. The `RazorGenerator.MsBuild` project includes a build task that needs to load these binaries - it's not meant to be references.

**Comment by wizzardmr42** [comment]
OK, I think I've been through everything outside the test project.  I'm not sure how to run the tests?  I just seem to get Assembly Resolve Exceptions when I try running them.  It seems to work for a couple of quick tests that I did manually on it with single file generators for MvcView and MvcHelper.  Probably a good idea if someone else tries to break it and maybe updates the test project (or at least give me a hand with getting it running?)  This is the first time I've contributed to open source, so not sure if I should submit a pull request now and we can sort the testing here or we test on my branch and make sure it is all stable before I submit it?

**Comment by CZEMacLeod** [comment]
How did you get on with this? I would be very interested in a version of this that could a) handle .vbhtml as well as .cshtml source files and b) generate .vb code instead of .vs code so it could be compiled into a VB based DLL.
I currently use an embedded resource virtual file provider and engine to embed razor files but would prefer to pre-compile them.

**Comment by pranavkm** [comment]
@wizzardmr42 sorry I missed your last comment. We can figure out the test in a follow up.

**Comment by wizzardmr42** [comment]
@CZEMacLeod I have submitted a pull request and dealt with a few minor issues that pranavkm pointed out.  The tests still need updating to handle VB.NET support, but someone else will need to do that.  I guess we'll watch and see what happens.

**Comment by davidebbo** [comment]
PR is #85. @pranavkm are you ok with it now after the changes? I think it's ok to start out without VB specific tests. The key is to make sure that it doesn't affect codegen for C#.

**Comment by pranavkm** [comment]
Yup, the PR looks good. I wanted to run it locally once to verify if the csharp codegen tests run as part of the build which should tell if if C# is affected in any way. I'll merge it in once I do that.

**Comment by CZEMacLeod** [comment]
@wizzardmr42 Does this update support both reading .vbhtml files and generating .vb code files? Does it support cross language scenarios? e.g. If my project is a VB DLL and has .cshtml views does it create .vb files to compile?

**Comment by wizzardmr42** [comment]
@CZEMacLeod I did it a few months ago so don't remember for sure but I think that it assumes that you want a .vb output for .vbhtml and a .cs for .cshtml.  It should be possible to change that though by checking for a .csproj or .vbproj file if you want to fork it

**Comment by davidebbo** [comment]
I don't think that makes sense. A .cshml file can contains actual user C# code, so there is no way you're going to generate VB out of it (and vice versa).

Bottom line is that when using RG, you need your view language and code file language to match.

**Comment by wizzardmr42** [comment]
Hadn't thought of that, although in theory, couldn't you translate it into CodeDOM objects and then back into the other language?  Seems like a lot of work though and not sure you'd want to put all that in the main branch

**Comment by davidebbo** [comment]
Nope, CodeDom has no support for that. The minute your CodeDom tree has any actual code snippets (as opposed to abstract blocks that are language agnostic), it is tied to that one language.

**Comment by CZEMacLeod** [comment]
I have the Roslyn codedom compilers enabled for my project - so I thought that the code blocks as well as the abstract blocks would be turned into expression trees (or whatever it is that is the in memory representation for Roslyn) so then could be written back in whichever language.
It isn't a huge issue for me to match razor language to the project file but I do have some of my current DLLs with a mix of both .vbhtml and .cshtml pages

**Comment by davidebbo** [comment]
Though I haven't worked with the Roslyn codedom providers, I'd be very surprised if they were this smart. I think the code generation is probably exactly the same. The only difference is how it compiles code once it's all generated.

**Comment by matvdl** [comment]
hello - I see that the code now supports VB.net - that is great news.

I have had a go but unfortunately I can't get it working?  I have added the RazorGenerator extension, created a standard vb class library, added the RazorGenerator template nugget package to the project. I then renamed the SampleTemplate.cshtml to vbhtml and No class file is generated for the SampleTemplate file.  When I right click on the file and select "Run Custom Tool"  I get the error "Cannot find customer tool 'RazorGenerator' on this system".

If I follow the exact same steps - but create a C# class library everything works perfectly.

I have attached a copy of the project if this helps - but it is very simple.  Hopefully I have not missed something really simple.
[TestRazorVB.zip](https://github.com/RazorGenerator/RazorGenerator/files/268051/TestRazorVB.zip)
